experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeBinder.ts

```diff
@@ -0,0 +1,930 @@
+ *
+ * @alpha
+ */
+export type BindPath = DownPath;
+
+/**
+ * @alpha
+ */
+export type BindingContext = DeleteBindingContext | InsertBindingContext | SetValueBindingContext;
+
+/**
+ * @alpha
+ */
+export type BindingContextQueue = BindingContext[];
+
+/**
+ * @alpha
+ */
+export const BindingType = {
+	Delete: "delete",
+	Insert: "insert",
+	SetValue: "setValue",
+	Invalidation: "invalidation",
+	Batch: "batch",
+} as const;
+
+/**
+ * @alpha
+ */
+export type BindingContextType = typeof BindingType[keyof typeof BindingType];
+
+/**
+ * @alpha
+ */
+export interface AbstractBindingContext {
+	readonly type: BindingContextType;
+}
+
+/**
+ * @alpha
+ */
+export interface DeleteBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.Delete;
+	readonly path: UpPath;
+	readonly count: number;
+}
+
+/**
+ * @alpha
+ */
+export interface InsertBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.Insert;
+	readonly path: UpPath;
+	readonly content: ProtoNodes;
+}
+
+/**
+ * @alpha
+ */
+export interface SetValueBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.SetValue;
+	readonly path: UpPath;
+	readonly value: TreeValue;
+}
+
+/**
+ * @alpha
+ */
+export interface InvalidationBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.Invalidation;
+}
+
+/**
+ * @alpha
+ */
+export interface BatchBindingContext extends AbstractBindingContext {
+	readonly type: typeof BindingType.Batch;
+	readonly events: BindingContext[];
+}
+
+abstract class AbstractPathVisitor implements PathVisitor {
+	protected readonly registeredPaths: Map<
+		BindingContextType,
+		Map<BindPath, Set<(...args: unknown[]) => any>>
+	> = new Map();
+	public constructor(protected readonly options: BinderOptions) {}
+	public abstract onDelete(path: UpPath, count: number): void;
+	public abstract onInsert(path: UpPath, content: ProtoNodes): void;
+	public abstract onSetValue(path: UpPath, value: TreeValue): void;
+
+	protected matchesPath(visitPath: BindPath | undefined, downPath: BindPath): boolean {
+		if (visitPath !== undefined) {
+			const matchPolicy = this.options.matchPolicy;
+			switch (matchPolicy) {
+				case "subtree":
+					if (this.pathsMatchSubtree(visitPath, downPath)) {
+						return true;
+					}
+					break;
+				case "path":
+					if (this.pathsMatchExact(visitPath, downPath)) {
+						return true;
+					}
+					break;
+				default:
+					unreachableCase(matchPolicy);
+			}
+		}
+		return false;
+	}
+	// TODO optimize loop to hash check
+	protected pathsMatchExact(current: BindPath, other: BindPath) {
+		if (current.length !== other.length) {
+			return false;
+		}
+		return this.pathsMatchSubtree(current, other);
+	}
+	protected pathsMatchSubtree(current: BindPath, other: BindPath) {
+		for (const [i, step] of current.entries()) {
+			if (
+				step.field !== other[i].field ||
+				(step.index !== undefined && step.index !== other[i].index)
```

Comment:
The asymmetry is intentional and tries to cover both scenarios:
- bindTree specifications which include an index, signaling the intention that the match (either `path` or `subtree`) should _consider_ it when filtering events
- bindTree specifications which do not specify an index,  signaling the intention that the match (either `path` or `subtree`) should _disregard_ the index value (as present in the path tree) when filtering events

Renaming `pathsMatchExact` to `pathsMatchPath` to avoid further confusion.

Reply:
I see. So leaving the index undefined, means that we want to match all indices.